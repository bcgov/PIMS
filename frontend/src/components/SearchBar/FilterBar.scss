@import '../../fonts.scss';
.search-bar {
  padding: 12px;
  background-color: #f2f2f2;
  .bar-item {
    .form-group {
      padding: 0;
      margin: 0;
    }
  }
  .vl {
    border-left: 6px solid #666666;
    height: 40px;
    margin-left: 1%;
    margin-right: 1%;
    border-width: 1px;
  }
  button.btn {
    width: 44px;
    height: 36px;
  }
}
