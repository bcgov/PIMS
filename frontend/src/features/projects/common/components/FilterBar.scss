@import '../../../../colors.scss';

.ProjectDisposeView {
  .filter-container {
    background-color: $filter-background-color;

    .filter-bar {
      align-items: center;
      padding: 1rem 1rem;

      .bar-item {
        max-width: 250px;
        .form-group {
          padding: 0;
          margin: 0;
        }
        .input-group {
          display: block;
        }
      }

      .btn-warning {
        color: white;
        font-weight: bold;
      }

      .form-control {
        border-radius: 0;
      }
    }
  }
}
