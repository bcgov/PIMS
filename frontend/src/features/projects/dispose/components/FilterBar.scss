@import '../../../../fonts.scss';

.ProjectDisposeView {
  .filter-container {
    background-color: #f2f2f2;

    .filter-bar {
      align-items: center;
      padding: 1rem 1rem;

      .bar-item {
        .form-group {
          padding: 0;
          margin: 0;
        }
        .input-group {
          display: block;
        }
      }

      .btn-warning {
        color: white;
        font-weight: bold;
      }

      .form-control {
        border-radius: 0;
      }
    }
  }
}
