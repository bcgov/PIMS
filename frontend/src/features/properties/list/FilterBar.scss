@import '../../../fonts.scss';

.filter-container {
  background-color: #f2f2f2;

  .filter-bar {
    align-items: center;
    padding: 1rem 0;

    .bar-item {
      .form-group {
        padding: 0;
        margin: 0;
      }
    }

    .btn-warning {
      color: white;
      font-weight: bold;
    }

    .form-control {
      border-radius: 0;
    }
  }
}
