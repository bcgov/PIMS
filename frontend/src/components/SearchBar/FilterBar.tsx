import './FilterBar.scss';

import React, { PropsWithChildren } from 'react';
import { Col } from 'react-bootstrap';
import { Formik } from 'formik';
import { Form } from 'components/common/form';
import ResetButton from 'components/common/form/ResetButton';
import SearchButton from 'components/common/form/SearchButton';
import PlusButton from 'components/common/form/PlusButton';

interface IProps<T extends object = {}> {
  initialValues: T;
  onChange: (value: any) => void;
  /** controls the className of the search button */
  searchClassName?: string;
  /** determine whether to hide the reset button */
  hideReset?: boolean;
  /** determine whether the plus button is needed for this filter bar */
  plusButton?: boolean;
  /** make a heading for this fitler bar */
  filterBarHeading?: string;
  /** if the filter bar is enabled to add a new entity control what happens on click */
  handleAdd?: (value: any) => void;
  /** the two probs below are for controlling the tool tip for the plust button */
  toolTipAddId?: string;
  toolTipAddText?: string;
  /** override formiks handleReset default for customized components */
  customReset?: () => void;
  /** custom component field name to clear/reset */
  customResetField?: string;
}

const FilterBar = <T extends object = {}>(props: PropsWithChildren<IProps<T>>) => {
  return (
    <Formik<T>
      initialValues={props.initialValues}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        props.onChange?.({ ...values });
        setSubmitting(false);
      }}
      onReset={(values, { setSubmitting }) => {
        setSubmitting(true);
        props.onChange?.({ ...values });
        setSubmitting(false);
      }}
    >
      {({ isSubmitting, handleReset, setFieldValue }) => (
        <Form>
          <Form.Row className="search-bar">
            <h3 className="filterBarHeading">{props.filterBarHeading}</h3>
            {props.children}
            <Col className="bar-item flex-grow-0">
              <SearchButton
                className={props.searchClassName ? props.searchClassName : 'bg-warning'}
                disabled={isSubmitting}
              />
            </Col>
            {!props.hideReset && (
              <Col className="bar-item flex-grow-0">
                <ResetButton
                  disabled={isSubmitting}
                  onClick={() => {
                    if (props.customReset && props.customResetField) {
                      props.customReset();
                      setFieldValue(props.customResetField, '');
                    } else {
                      handleReset();
                    }
                  }}
                />
              </Col>
            )}
            {props.plusButton && (
              <>
                <div className="vl"></div>
                <Col className="bar-item flex-grow-0 plus-button">
                  <PlusButton
                    onClick={props.handleAdd}
                    toolId={props.toolTipAddId!}
                    toolText={props.toolTipAddText!}
                  />
                </Col>
              </>
            )}
          </Form.Row>
        </Form>
      )}
    </Formik>
  );
};

export default FilterBar;
