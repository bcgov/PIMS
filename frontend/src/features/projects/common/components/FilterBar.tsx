import './FilterBar.scss';

import React from 'react';
import { Col } from 'react-bootstrap';
import { Formik, useFormikContext } from 'formik';
import { Form, Select, InputGroup, Input } from 'components/common/form';
import ResetButton from 'components/common/form/ResetButton';
import SearchButton from 'components/common/form/SearchButton';
import { useCodeLookups } from 'hooks/useLookupCodes';
import * as API from 'constants/API';
import { Classifications } from 'constants/classifications';

const SearchBar: React.FC = () => {
  const state: { placeholders: Record<string, string> } = {
    placeholders: {
      address: 'Enter an address or city',
    },
  };

  // access the form context values, no need to pass props
  const {
    values: { searchBy },
  } = useFormikContext<IFilterBarState>();
  const desc = state.placeholders[searchBy] || '';

  return (
    <InputGroup
      fast={false}
      formikProps={null as any}
      field={searchBy}
      placeholder={desc}
    ></InputGroup>
  );
};

export interface IFilterBarState {
  searchBy: string;
  pid: string;
  address: string;
  administrativeArea: string;
  projectNumber: string;
  agencies: string;
  classificationId: string;
  minLotSize: string;
  maxLotSize: string;
}

type FilterBarProps = {
  onChange: (value: IFilterBarState) => void;
};

/**
 * Filter bar for the Property List view
 */
const FilterBar: React.FC<FilterBarProps> = ({ onChange }) => {
  const lookupCode = useCodeLookups();
  //restrict available agencies to user agencies.
  const agencies = lookupCode.getOptionsByType(API.AGENCY_CODE_SET_NAME);
  const classifications = lookupCode.getPropertyClassificationOptions(
    c =>
      +c.value !== Classifications.Demolished &&
      +c.value !== Classifications.Disposed &&
      +c.value !== Classifications.Subdivided,
  );

  return (
    <Formik<IFilterBarState>
      initialValues={{
        searchBy: 'address',
        pid: '',
        address: '',
        administrativeArea: '',
        projectNumber: '',
        agencies: '',
        classificationId: '',
        minLotSize: '',
        maxLotSize: '',
      }}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        onChange?.({ ...values });
        setSubmitting(false);
      }}
      onReset={(values, { setSubmitting }) => {
        setSubmitting(true);
        onChange?.({ ...values });
        setSubmitting(false);
      }}
    >
      {({ isSubmitting, handleReset }) => (
        <Form>
          <Form.Row className="filter-bar">
            <Col className="bar-item">
              <Input field="pid" placeholder="Enter PID or PIN" />
            </Col>
            <Col className="bar-item">
              <SearchBar />
            </Col>
            <Col className="bar-item">
              <Select field="agencies" placeholder="Enter an Agency" options={agencies} />
            </Col>
            <Col className="bar-item">
              <Select
                field="classificationId"
                placeholder="Classification"
                options={classifications}
              />
            </Col>
            <Col className="bar-item flex-grow-0">
              <SearchButton disabled={isSubmitting} />
            </Col>
            <Col className="bar-item flex-grow-0">
              <ResetButton disabled={isSubmitting} onClick={handleReset} />
            </Col>
          </Form.Row>
        </Form>
      )}
    </Formik>
  );
};

export default FilterBar;
