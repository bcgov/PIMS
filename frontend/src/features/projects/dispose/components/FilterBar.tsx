import './FilterBar.scss';

import React from 'react';
import { Col } from 'react-bootstrap';
import { Formik, useFormikContext } from 'formik';
import { ILookupCode } from 'actions/lookupActions';
import {
  Form,
  Select,
  SelectOption,
  Button,
  ButtonProps,
  InputGroup,
} from 'components/common/form';
import { FaUndo, FaSearch } from 'react-icons/fa';
import _ from 'lodash';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';

const SearchButton: React.FC<ButtonProps> = ({ ...props }) => {
  return <Button type="submit" className="bg-warning" {...props} icon={<FaSearch size={20} />} />;
};

const ResetButton: React.FC<ButtonProps> = ({ ...props }) => {
  return (
    <Button type="reset" variant="outline-primary" {...props} icon={<FaUndo size={20} />}>
      Reset
    </Button>
  );
};

const SearchBar: React.FC = () => {
  const state: { placeholders: Record<string, string> } = {
    placeholders: {
      address: 'Enter an address or city',
    },
  };

  // access the form context values, no need to pass props
  const {
    values: { searchBy },
  } = useFormikContext<IFilterBarState>();
  const desc = state.placeholders[searchBy] || '';

  return (
    <InputGroup
      fast={false}
      formikProps={null as any}
      field={searchBy}
      placeholder={desc}
    ></InputGroup>
  );
};

export interface IFilterBarState {
  searchBy: string;
  address: string;
  municipality: string;
  projectNumber: string;
  agencies: string;
  classificationId: string;
  minLotSize: string;
  maxLotSize: string;
}

type FilterBarProps = {
  agencyLookupCodes: ILookupCode[];
  propertyClassifications: ILookupCode[];
  onChange: (value: IFilterBarState) => void;
};

/**
 * Filter bar for the Property List view
 */
const FilterBar: React.FC<FilterBarProps> = ({
  agencyLookupCodes,
  propertyClassifications,
  onChange,
}) => {
  const mapLookupCode = (code: ILookupCode): SelectOption => ({
    label: code.name,
    value: code.id.toString(),
  });
  const keycloak = useKeycloakWrapper();
  //restrict available agencies to user agencies.
  const agencies = _.filter(
    (agencyLookupCodes ?? []).map(c => mapLookupCode(c)),
    agency => keycloak.hasAgency(parseInt(agency.value as string)),
  );
  const classifications = (propertyClassifications ?? []).map(c => mapLookupCode(c));

  return (
    <Formik<IFilterBarState>
      initialValues={{
        searchBy: 'address',
        address: '',
        municipality: '',
        projectNumber: '',
        agencies: '',
        classificationId: '',
        minLotSize: '',
        maxLotSize: '',
      }}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        onChange?.({ ...values });
        setSubmitting(false);
      }}
      onReset={(values, { setSubmitting }) => {
        setSubmitting(true);
        onChange?.({ ...values });
        setSubmitting(false);
      }}
    >
      {({ isSubmitting, handleReset }) => (
        <Form>
          <Form.Row className="filter-bar">
            <Col className="bar-item">
              <SearchBar />
            </Col>
            <Col className="bar-item">
              <Select field="agencies" placeholder="Enter an Agency" options={agencies} />
            </Col>
            <Col className="bar-item">
              <Select
                field="classificationId"
                placeholder="Classification"
                options={classifications}
              />
            </Col>
            <Col className="bar-item flex-grow-0">
              <SearchButton disabled={isSubmitting} />
            </Col>
            <Col className="bar-item flex-grow-0">
              <ResetButton disabled={isSubmitting} onClick={handleReset} />
            </Col>
          </Form.Row>
        </Form>
      )}
    </Formik>
  );
};

export default FilterBar;
