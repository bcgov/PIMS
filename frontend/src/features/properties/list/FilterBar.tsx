import './FilterBar.scss';

import React, { useMemo } from 'react';
import { Col } from 'react-bootstrap';
import { Formik, useFormikContext } from 'formik';
import { ILookupCode } from 'actions/lookupActions';
import { Form, Select, SelectOption, InputGroup, Input } from 'components/common/form';
import ResetButton from 'components/common/form/ResetButton';
import SearchButton from 'components/common/form/SearchButton';
import { BasePropertyFilter } from 'components/common/interfaces';
import { ParentGroupedFilter } from 'components/SearchBar/ParentGroupedFilter';

const SearchBar: React.FC = () => {
  const state: { options: any[]; placeholders: Record<string, string> } = {
    options: [
      { label: 'Address', value: 'address' },
      { label: 'PID/PIN', value: 'pid' },
      { label: 'Location', value: 'administrativeArea' },
      { label: 'RAEG or SPP No.', value: 'projectNumber' },
    ],
    placeholders: {
      address: 'Enter an address or location',
      pid: 'Enter a PID or PIN',
      administrativeArea: 'Enter a location',
      projectNumber: 'Enter an SPP/RAEG number',
    },
  };

  // access the form context values, no need to pass props
  const {
    values: { searchBy },
    setFieldValue,
  } = useFormikContext<IFilterBarState>();
  const desc = state.placeholders[searchBy] || '';

  const reset = () => {
    setFieldValue('address', '');
    setFieldValue('administrativeArea', '');
    setFieldValue('projectNumber', '');
  };

  return (
    <InputGroup
      fast={false}
      formikProps={null as any}
      prepend={<Select field="searchBy" options={state.options} onChange={reset} />}
      field={searchBy}
      placeholder={desc}
    ></InputGroup>
  );
};

export interface IFilterBarState extends BasePropertyFilter {
  searchBy: string;
  pid: string;
  address: string;
  administrativeArea: string;
  projectNumber: string;
  agencies: string;
  classificationId: string;
  minLotSize: string;
  maxLotSize: string;
  parcelId?: string;
  propertyType?: PropertyTypes;
}

export enum PropertyTypes {
  Land = 'Land',
  Building = 'Building',
}

type FilterBarProps = {
  agencyLookupCodes: ILookupCode[];
  propertyClassifications: ILookupCode[];
  filter?: IFilterBarState;
  onChange: (value: IFilterBarState) => void;
};

const defaultFilterValues: IFilterBarState = {
  searchBy: 'address',
  pid: '',
  address: '',
  administrativeArea: '',
  projectNumber: '',
  agencies: '',
  classificationId: '',
  minLotSize: '',
  maxLotSize: '',
  parcelId: '',
  propertyType: PropertyTypes.Land,
};

/**
 * Filter bar for the Property List view
 */
const FilterBar: React.FC<FilterBarProps> = ({
  agencyLookupCodes,
  propertyClassifications,
  filter,
  onChange,
}) => {
  const mapLookupCode = (code: ILookupCode): SelectOption => ({
    label: code.name,
    value: code.id.toString(),
    code: code.code,
    parentId: code.parentId,
    parent: agencyLookupCodes.find(x => x.id.toString() === code.parentId?.toString())?.name,
  });

  const agencies = (agencyLookupCodes ?? []).map(c => mapLookupCode(c));
  const initialValues = useMemo(() => {
    const values = { ...defaultFilterValues, ...filter };
    if (typeof values.agencies === 'string') {
      const agency = agencies.find(x => x.value.toString() === values.agencies.toString()) as any;
      if (agency) {
        values.agencies = agency;
      }
    }
    return values;
  }, [agencies, filter]);

  const classifications = (propertyClassifications ?? [])
    .filter(i => !!i.isVisible)
    .map(c => mapLookupCode(c));

  return (
    <Formik<IFilterBarState>
      initialValues={{ ...initialValues }}
      enableReinitialize
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        onChange?.({ ...values, agencies: (values.agencies as any)?.value });
        setSubmitting(false);
      }}
    >
      {({ isSubmitting, handleReset }) => (
        <Form>
          <Form.Row className="filter-bar">
            <Col className="bar-item">
              <SearchBar />
            </Col>
            <Col className="bar-item">
              <ParentGroupedFilter
                name="agencies"
                options={agencies}
                className="map-filter-typeahead"
                filterBy={['code', 'label', 'parent']}
                placeholder="Enter an Agency"
                inputSize="large"
              />
            </Col>
            <Col className="bar-item">
              <Select
                field="classificationId"
                placeholder="Classification"
                options={classifications}
              />
            </Col>
            <Col className="bar-item d-flex align-items-center">
              <Input field="minLotSize" placeholder="Min Lot Size" />
              <span className="mx-2">-</span>
              <Input field="maxLotSize" placeholder="Max Lot Size" />
            </Col>
            <Col className="bar-item flex-grow-0">
              <SearchButton disabled={isSubmitting} />
            </Col>
            <Col className="bar-item flex-grow-0">
              <ResetButton
                disabled={isSubmitting}
                onClick={() => onChange?.({ ...defaultFilterValues })}
              />
            </Col>
          </Form.Row>
        </Form>
      )}
    </Formik>
  );
};

export default FilterBar;
