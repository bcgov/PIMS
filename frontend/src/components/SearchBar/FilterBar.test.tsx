import Adapter from '@cfaester/enzyme-adapter-react-18';
import Enzyme from 'enzyme';
import React from 'react';
import { create } from 'react-test-renderer';

import FilterBar from './FilterBar';

Enzyme.configure({ adapter: new Adapter() });

const componentRender = () => {
  const component = create(
    <div>
      <FilterBar initialValues={{ username: 'test', firstName: 'user' }} onChange={() => {}} />
    </div>,
  );
  return component;
};

describe('Filter Bar', () => {
  it('Snapshot matches', () => {
    const component = componentRender();
    expect(component.toJSON()).toMatchSnapshot();
  });
});
