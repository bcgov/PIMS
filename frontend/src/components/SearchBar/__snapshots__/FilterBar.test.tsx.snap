// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Bar Snapshot matches 1`] = `
<div>
  <form
    className=""
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="search-bar form-row"
    >
      <div
        className="bar-item flex-grow-0 col"
      >
        <button
          className="Button Button--icon-only btn btn-warning"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="submit"
        >
          <div
            className="Button__icon"
          >
            <svg
              fill="currentColor"
              height={20}
              size={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        className="bar-item flex-grow-0 col"
      >
        <button
          className="Button Button--icon-only btn btn-secondary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="reset"
        >
          <div
            className="Button__icon"
          >
            <svg
              fill="currentColor"
              height={20}
              size={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </form>
</div>
`;
